packages/api/src/agent.ts

```diff
@@ -159,23 +159,45 @@ export class AtpAgent {
     try {
       this.session = session
       const res = await this.api.com.atproto.server.getSession()
-      if (!res.success || res.data.did !== this.session.did) {
-        throw new Error('Invalid session')
+      if (res.data.did !== this.session.did) {
+        throw new XRPCError(
+          ResponseType.InvalidRequest,
+          'Invalid session',
+          'InvalidDID',
+        )
       }
       this.session.email = res.data.email
       this.session.handle = res.data.handle
       this.session.emailConfirmed = res.data.emailConfirmed
       this._updateApiEndpoint(res.data.didDoc)
+      this._persistSession?.('update', this.session)
       return res
     } catch (e) {
       this.session = undefined
-      throw e
-    } finally {
-      if (this.session) {
-        this._persistSession?.('create', this.session)
+
+      if (e instanceof XRPCError && e.error) {
+        /*
+         * `ExpiredToken` and `InvalidToken` are handled in
+         * `this_refreshSession`, and emit an `expired` event there.
+         *
+         * `AuthMissing` is a respose from `getSession`, and happens AFTER the
+         * initial `expired` event is sent. If we handle that here, we'll
+         * double-emit `expired`.
+         *
+         * `InvalidDID` is emit above, and should be handled here.
+         *
+         * Everything else is unexpected.
+         */
+        if (['InvalidDID'].includes(e.error)) {
```

Comment:
I'm a bit confused what circumstance this happens in?

Reply:
This is just handling the logic that was already here, where a did doesn't match the requested session did. I honestly don't know why this would happen.

Reply:
ah cool cool 👍

Reply:
and i guess if that happens, then somethings gone wonky & the appropriate action would be to just log the user out - hence the `expired`